firedrake/assemble.py

```diff
@@ -214,11 +216,23 @@ def _assemble(f, tensor=None, bcs=None, form_compiler_parameters=None,
         kernels = tsfc_interface.compile_form(f, "form", parameters=form_compiler_parameters, inverse=inverse)
         integral_types = [integral.integral_type() for integral in f.integrals()]
 
+        if bcs is not None:
+            for bc in bcs:
+                if isinstance(bc, FormBC):
+                    integral_types += [integral.integral_type() for integral in bc.f.integrals()]
```

Comment:
Could we have a `FormBC` provide an `integrals()` method which returns the integrals, with a `DirichletBC` returning an empty list? Then we don't need this instance check.